packages/router/utils.ts

```diff
@@ -1411,3 +1431,37 @@ export function isRouteErrorResponse(error: any): error is ErrorResponse {
     "data" in error
   );
 }
+
+/**
+ * Context object passed through middleware functions and into action/loaders.
+ *
+ * Supports only key/value for now, eventually will be enhanced
+ */
+export interface MiddlewareContext<T = unknown> {
+  get(key: MiddlewareContextInstance<T>): T;
+  set(key: MiddlewareContextInstance<T>, value: T): void;
+  next: () => DataFunctionReturnValue;
+}
+
+/**
+ * Generic class to "hold" a default middleware value and the generic type so
+ * we can enforce typings on middleware.get/set
+ */
+export class MiddlewareContextInstance<T> {
+  private defaultValue: T | null;
+
+  constructor(defaultValue?: T | null) {
+    this.defaultValue =
+      typeof defaultValue !== "undefined" ? defaultValue : null;
+  }
+
+  getDefaultValue() {
+    return this.defaultValue;
+  }
+}
+
+export function createMiddlewareContext<T extends unknown>(
+  defaultValue: T
+): MiddlewareContextInstance<T> {
+  return new MiddlewareContextInstance<T>(defaultValue);
+}
```

Comment:
@pcattori Definitely want to get your eyes on this before merging - this all "works" but no clue if it's "right" ;)

Reply:
Try second generic for `defaultValue` type